tun/src/unix/linux/mod.rs

```diff
@@ -75,6 +75,24 @@ impl TunInterface {
         let addr = unsafe { *(&iff.ifr_ifru.ifru_addr as *const _ as *const sys::sockaddr_in) };
         Ipv4Addr::from(u32::from_be(addr.sin_addr.s_addr))
     }
+    
+    #[throws]
+    pub fn set_ipv6_addr(&self, addr: Ipv6Addr) {
+        let addr = SockAddr::from(SocketAddrV6::new(addr, 0, 0, 0));
+        
+        let mut iff = self.ifreq()?;
+        iff.ifr_ifru.ifru_addr = unsafe { *addr.as_ptr() };
```

Comment:
I did not get a chance to write this in the issue description, but there is actually a separate request structure for IPv6 addresses:

[ifreq](https://github.com/rust-lang/libc/blob/5744793b48ddd451ed0690e486845f60280150a3/src/unix/linux_like/linux/mod.rs#L776-L800)
[in6_ifreq](https://github.com/rust-lang/libc/blob/5744793b48ddd451ed0690e486845f60280150a3/src/unix/linux_like/linux/mod.rs#L619-L623)

So what you'll want to do is make a new method, `fn in6_ifreq(&self) -> sys::in6_ifreq`, based off of the `ifreq` method. This new method should set `ifr6_ifindex` on a new `in6_ifreq` struct, as described [here](https://man7.org/linux/man-pages/man7/netdevice.7.html)

Once you have this structure, then you need to set and get the `ifr6_addr.s6_addr` field, which hass the type `[u8; 16]`. You can get it from an Ipv6Addr using [`octets`](https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html#method.octets) and go in reverse using [`from`](https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html#impl-From%3C%5Bu8;+16%5D%3E-for-Ipv6Addr).

You will also need two more things:

1. A new `ioctl` in `sys.rs`, because the type is different now:
```rust
ioctl_write_ptr_bad!(if_set_addr6, libc::SIOCSIFADDR, libc::in6_ifreq); // the `if_set_addr6` part is up to you
// getting an ipv6 addr also needs an ioctl
```
2. A new `perform6` method that models the `perform` method, but using `Domain::IPV6` instead of `Domain::IPV4`

Reply:
Thanks so much @conradev! Would I set the `ifr6_addr.s6_addr` field within `fn in6_ifreq(&self)` or within `fn set_ipv6_addr(&self, addr: Ipv6Addr)`?

Reply:
Within `fn set_ipv6_addr(&self, addr: Ipv6Addr)`, because you'll want to use the passed in address